src/lib/snark_work_lib/metrics.ml

```diff
@@ -0,0 +1,180 @@
+           with $zkapp_command_count zkapp_command and \
+           $proof_zkapp_command_count proof zkapp_command"
+      }]
+
+type Structured_log_events.t +=
+  | Sub_zkapp_snark_generated of
+      { elapsed : Time_span_with_json.t; kind : [ `Merge | `Segment ] }
+  [@@deriving
+    register_event
+      { msg = "Sub-zkApp SNARK generated in $elapsed of type $kind" }]
+
+let emit_single_metrics ~(single_spec : _ Single_spec.Poly.t)
+    ~data:
+      ({ data = elapsed; _ } :
+        (Core.Time.Stable.Span.V1.t, _) Proof_carrying_data.t ) =
+  match single_spec with
+  | Merge (_, _, _) ->
+      Perf_histograms.add_span ~name:"snark_worker_merge_time" elapsed ;
+      Mina_metrics.(
+        Cryptography.Snark_work_histogram.observe
+          Cryptography.snark_work_merge_time_sec (Time.Span.to_sec elapsed)) ;
+      Merge_snark_generated { elapsed }
+  | Transition (_, ({ transaction; _ } : Transaction_witness.Stable.Latest.t))
+    ->
+      Perf_histograms.add_span ~name:"snark_worker_transition_time" elapsed ;
+      let transaction_type, zkapp_command_count, proof_zkapp_command_count =
+        match transaction with
+        | Mina_transaction.Transaction.Command
+            (Mina_base.User_command.Zkapp_command zkapp_command) ->
+            (* WARN: now this is dead code, if we're using new protocol between
+               snark coordinator and workers *)
+            let parties_count = List.length zkapp_command.account_updates in
+            let proof_parties_count_init =
+              match
+                (Mina_base.Account_update.of_fee_payer
+                   zkapp_command.Mina_base.Zkapp_command.Poly.fee_payer )
+                  .authorization
+              with
+              | Proof _ ->
+                  1
+              | _ ->
+                  0
+            in
+            let proof_parties_count =
+              Mina_base.Zkapp_command.Call_forest.fold
+                zkapp_command.account_updates ~init:proof_parties_count_init
+                ~f:(fun cnt account_update ->
+                  if
+                    Mina_base.Control.(
+                      Tag.equal Proof
+                        (tag
+                           account_update
+                             .Mina_base.Account_update.Poly.authorization ))
+                  then cnt + 1
+                  else cnt )
+            in
+            Mina_metrics.(
+              Cryptography.(
+                Counter.inc snark_work_zkapp_base_time_sec
+                  (Time.Span.to_sec elapsed) ;
+                Counter.inc_one snark_work_zkapp_base_submissions ;
+                Counter.inc zkapp_transaction_length
+                  (Float.of_int parties_count) ;
+                Counter.inc zkapp_proof_updates
+                  (Float.of_int proof_parties_count))) ;
+            (`Zkapp_command, parties_count, proof_parties_count)
+        | Mina_transaction.Transaction.Command
```

Comment:
Let's unite the three cases below:

Note, you don't need to add prefix `Mina_base.User_command` to all cases, only to the first one, for the rest compiler would already know the type.

```ocaml
| Command (Signed_command _) | Coinbase _ | Fee_transfer _ ->
    Mina_metrics.(
              Counter.inc Cryptography.snark_work_non_zkapp_base_time_sec
                (Time.Span.to_sec elapsed)) ;
            (1, 0)
```

And let `transaction_type` be chosen via a separate match or via a call to a function defined separately.